File: android/src/main/kotlin/id/co/rfbnr/flutter_indocard_ocr/ModuleOCR/preprocessing/PreprocessingPipeline.kt
package id.co.rfbnr.flutter_indocard_ocr.ModuleOCR.preprocessing

import android.graphics.Bitmap
import android.util.Log

/**
 * Preprocessing Pipeline that orchestrates multiple preprocessors.
 * Executes preprocessors in sequence and handles errors gracefully.
 * 
 * Usage:
 * ```kotlin
 * val pipeline = PreprocessingPipeline.createDefault()
 * val result = pipeline.execute(bitmap)
 * val processedBitmap = result.getBitmapOrFallback(originalBitmap)
 * ```
 */
class PreprocessingPipeline(
    private val preprocessors: List<ImagePreprocessor>
) {

    companion object {
        private const val TAG = "PreprocessingPipeline"

        /**
         * Create default pipeline with card detection and image enhancement.
         */
        fun createDefault(): PreprocessingPipeline {
            return PreprocessingPipeline(
                listOf(
                    CardDetector(),
                    ImageEnhancer.createForOCR()
                )
            )
        }

        /**
         * Create pipeline with only image enhancement (no card detection).
         */
        fun enhanceOnly(): PreprocessingPipeline {
            return PreprocessingPipeline(
                listOf(ImageEnhancer.createForOCR())
            )
        }

        /**
         * Create pipeline with only grayscale conversion.
         */
        fun grayscaleOnly(): PreprocessingPipeline {
            return PreprocessingPipeline(
                listOf(ImageEnhancer.grayscaleOnly())
            )
        }
    }

    /**
     * Execute all preprocessors in sequence.
     * If any preprocessor fails, continues with the previous result.
     * 
     * @param bitmap The input bitmap to process
     * @return PreprocessResult with final processed bitmap or error
     */
    suspend fun execute(bitmap: Bitmap): PreprocessResult {
        var currentBitmap = bitmap
        var hasError = false
        val errors = mutableListOf<String>()

        Log.d(TAG, "Starting preprocessing pipeline with ${preprocessors.size} preprocessors")

        for (preprocessor in preprocessors) {
            try {
                Log.d(TAG, "Executing preprocessor: ${preprocessor.name}")
                
                when (val result = preprocessor.process(currentBitmap)) {
                    is PreprocessResult.Success -> {
                        currentBitmap = result.bitmap
                        Log.d(TAG, "${preprocessor.name} completed successfully")
                    }
                    is PreprocessResult.Error -> {
                        Log.w(TAG, "${preprocessor.name} failed: ${result.message}")
                        errors.add("${preprocessor.name}: ${result.message}")
                        hasError = true
                        // Continue with current bitmap (fallback behavior)
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Exception in ${preprocessor.name}: ${e.message}")
                errors.add("${preprocessor.name}: ${e.message}")
                hasError = true
                // Continue with current bitmap
            }
        }

        Log.d(TAG, "Preprocessing pipeline completed. Has errors: $hasError")

        return if (hasError && errors.isNotEmpty()) {
            // Return success with warnings - we still have a processed image
            PreprocessResult.Success(currentBitmap)
        } else {
            PreprocessResult.Success(currentBitmap)
        }
    }

    /**
     * Get list of preprocessor names in this pipeline.
     */
    fun getPreprocessorNames(): List<String> {
        return preprocessors.map { it.name }
    }

    /**
     * Release resources held by preprocessors.
     */
    fun close() {
        preprocessors.forEach { preprocessor ->
            if (preprocessor is CardDetector) {
                preprocessor.close()
            }
        }
    }
}

File: ios/Classes/ModuleOCR/preprocessing/PreprocessingPipeline.swift
import UIKit

/**
 * Preprocessing Pipeline that orchestrates multiple preprocessors.
 * Executes preprocessors in sequence and handles errors gracefully.
 *
 * Usage:
 * ```swift
 * let pipeline = PreprocessingPipeline.createDefault()
 * pipeline.execute(image) { result in
 *     let processedImage = result.getImageOrFallback(originalImage)
 *     // Use processedImage for OCR
 * }
 * ```
 */
class PreprocessingPipeline {
    
    private let preprocessors: [ImagePreprocessor]
    
    // MARK: - Initialization
    
    init(preprocessors: [ImagePreprocessor]) {
        self.preprocessors = preprocessors
    }
    
    /// Create default pipeline with card detection and image enhancement.
    static func createDefault() -> PreprocessingPipeline {
        return PreprocessingPipeline(preprocessors: [
            VisionCardDetector(),
            ImageEnhancer.forOCR()
        ])
    }
    
    /// Create pipeline with only image enhancement (no card detection).
    static func enhanceOnly() -> PreprocessingPipeline {
        return PreprocessingPipeline(preprocessors: [
            ImageEnhancer.forOCR()
        ])
    }
    
    /// Create pipeline with only grayscale conversion.
    static func grayscaleOnly() -> PreprocessingPipeline {
        return PreprocessingPipeline(preprocessors: [
            ImageEnhancer.grayscaleOnly()
        ])
    }
    
    // MARK: - Execution
    
    /**
     * Execute all preprocessors in sequence.
     * If any preprocessor fails, continues with the previous result.
     *
     * - Parameter image: The input image to process
     * - Parameter completion: Callback with final processed image or error
     */
    func execute(_ image: UIImage, completion: @escaping (PreprocessResult) -> Void) {
        print("[PreprocessingPipeline] Starting pipeline with \(preprocessors.count) preprocessors")
        
        executeRecursive(
            image: image,
            index: 0,
            errors: [],
            completion: completion
        )
    }
    
    /**
     * Recursively execute preprocessors in sequence.
     */
    private func executeRecursive(
        image: UIImage,
        index: Int,
        errors: [String],
        completion: @escaping (PreprocessResult) -> Void
    ) {
        // Base case: all preprocessors executed
        guard index < preprocessors.count else {
            print("[PreprocessingPipeline] Pipeline completed. Errors: \(errors.isEmpty ? "none" : errors.joined(separator: ", "))")
            completion(.success(image))
            return
        }
        
        let preprocessor = preprocessors[index]
        print("[PreprocessingPipeline] Executing: \(preprocessor.name)")
        
        preprocessor.process(image) { [weak self] result in
            guard let self = self else {
                completion(.error("Pipeline was deallocated", originalImage: image))
                return
            }
            
            var updatedErrors = errors
            let nextImage: UIImage
            
            switch result {
            case .success(let processedImage):
                print("[PreprocessingPipeline] \(preprocessor.name) completed successfully")
                nextImage = processedImage
                
            case .error(let message, let originalImage):
                print("[PreprocessingPipeline] \(preprocessor.name) failed: \(message)")
                updatedErrors.append("\(preprocessor.name): \(message)")
                // Continue with previous image (fallback)
                nextImage = originalImage ?? image
            }
            
            // Continue to next preprocessor
            self.executeRecursive(
                image: nextImage,
                index: index + 1,
                errors: updatedErrors,
                completion: completion
            )
        }
    }
    
    // MARK: - Utility
    
    /// Get list of preprocessor names in this pipeline.
    func getPreprocessorNames() -> [String] {
        return preprocessors.map { $0.name }
    }
}
